Yoast/wordpress-seo
# Introduce network admin-specific admin bar menu

## Summary

This PR can be summarized in the following changelog entry:

* Introduces a network admin-specific admin bar menu.

## Relevant technical choices:

* This PR ties in with the changes made in #9688 and therefore should be merged into the branch for the parent issue #9447. That branch will be merged into trunk when #9680, #9688, #9689, and #9875 have been completed.
* A `wpseo_network_admin_bar_menu()` function is introduced that takes care of rendering the admin bar content for the network admin.
    * It always contains the "Keyword Research" submenu.
    * It contains an "SEO Settings" submenu if the current user has access to the network settings.
    * The network admin menu at this point is always displayed, as there is no network setting to disable the network admin menu, and it would not make sense to rely on the site-specific setting for that. If this is a priority to have, we can integrate a network setting for that.
* The existing `wpseo_admin_bar_menu()` function now only operates in the frontend and the site admin.
* Some functionality previously part of `wpseo_admin_bar_menu()` has been outsourced into separate helper functions to be reusable:
    * `wpseo_adminbar_title()` returns the title markup (the logo icon with a11y text).
    * `wpseo_adminbar_add_keyword_research_submenu()` adds the "Keyword Research" submenu, and you can pass the parent menu to append it to as a function parameter. Furthermore I adjusted this function so that it now only includes query strings for the URLs if a focus keyword is present (i.e. it now generates URLs such as _http://tools.seobook.com/keyword-tools/seobook/_ instead of _http://tools.seobook.com/keyword-tools/seobook/?keyword=_ in those cases).
* `wpseo_admin_bar_style()` has been adjusted to load the styles as necessary.

**UPDATE:** I completely rewrote the PR, introducing a dedicated admin bar class as suggested. Some details on the implementation:

* `WPSEO_Admin_Bar_Menu` implements the `WPSEO_WordPress_Integration` interface for adding hooks. Its core are a `add_menu()` method and a `enqueue_assets()` method.
* The class handles the admin bar menu for all areas (frontend, site admin, network admin). This is similar how WordPress core does it, and it makes sense due to the significant overlap in functionality.
* By passing in the `WP_Admin_Bar` instance into each respective method, we can now easily unit test those.
* To make critical typos less likely, reusable identifiers for the admin bar menu and submenus have been added as class constants.
* I also refactored the `WPSEO_Admin_Menu` and `WPSEO_Network_Admin_Menu` class to be more aligned with each other. In order to make this straightforward, I introduced an abstract `WPSEO_Base_Admin_Menu` class with common functionality. The main reason I did this is so that the admin bar can automatically access the identifiers, titles and required capabilities for each settings menu page and submenu pages. This is an improvement over how the admin bar previously simply redeclared all the menu items, being prone to errors and the titles or URLs getting out-of-sync. With the current approach, it is ensured that admin menu and admin bar are always consistent.
* All the admin bar utility functions that were previously part of `wpseo-non-ajax-functions.php` were migrated to become protected methods of the new admin bar class. I implemented several of them in a more reusable way, where you pass in a certain post or term, to ensure responsibilities for each method are more obvious.
* All admin bar functions part of `wpseo-non-ajax-functions.php` have been deprecated.

## Test instructions

This PR can be tested by following these steps:

* Setup a multisite environment, and make sure your user has network administrator access (should be by default).
* Go to the network admin.
* Activate the plugin network-wide.
* Ensure that the Yoast SEO admin bar menu is present and that it contains the Keyword Research menu as well as the SEO Settings menu, with correct links to the three network admin screens of the plugin.
* Go back to the site admin.
* Ensure that the Yoast SEO admin bar menu still appears exactly the same as before. Pay particular attention to whether the same submenu items are present and to whether the logo icon is visible (i.e. the styles are correctly loaded).
* Go into the frontend.
* Check if the Yoast SEO admin bar menu still appears correctly. You don't need to be as precise here for the individual menu items, since in the frontend the same logic as for the site admin is used.
* Ensure that when editing or viewing a post with a focus keyword set, the Google Trends and SEO Book links in the Keyword Research submenu take that focus keyword into account. Also ensure that when no focus keyword is available in the current context (i.e. when not editing or viewing a post), these links should not contain empty query arguments or anchors.

## Quality assurance


Fixes #9875
